src/V3Const.cpp

```diff
@@ -76,6 +76,418 @@ class ConstVarFindVisitor final : public AstNVisitor {
+            if (!srcp) srcp = m_refp;
+            const int width = widthOfRef(m_refp);
+
+            if (hasConstantResult())
+                return new AstConst{fl,
+                                    V3Number{srcp, width, static_cast<vluint32_t>(m_constResult)}};
+
+            AstConst* maskValuep = new AstConst{fl, V3Number{srcp, width, 0}};
+            maskValuep->num().opBitsNonX(m_bitPolarity);  // 'x' -> 0, 0->1, 1->1
+            // Let AstConst be in lhs as it is the common convention
+            AstAnd* maskedp = new AstAnd{fl, maskValuep, srcp->cloneTree(false)};
+            AstNode* resultp;
+            if (m_parent->isXorTree()) {
+                resultp = new AstRedXor{fl, maskedp};
+                resultp->dtypep()->widthForce(width, 1);
+            } else {
+                AstConst* compValuep = maskValuep->cloneTree(false);
+                compValuep->num().opBitsOne(m_bitPolarity);  // 'x'->0, 0->0, 1->1
+                if (m_parent->isAndTree()) {
+                    resultp = new AstEq{fl, compValuep, maskedp};
+                } else {  // Or
+                    compValuep->num().opXor(V3Number{compValuep->num()}, maskValuep->num());
+                    resultp = new AstNeq{fl, compValuep, maskedp};
+                }
+            }
+            return resultp;
+        }
+
+        // CONSTRUCTORS
+        VarInfo(ConstBitOpTreeVisitor* parent, AstVarRef* refp)
+            : m_parent(parent)
+            , m_refp(refp)
+            , m_bitPolarity(refp, widthOfRef(refp)) {
+            m_bitPolarity.setAllBitsX();
+        }
+    };
+
+    // MEMBERS
+    bool m_failed = false;
+    bool m_polarity = true;  // Flip when AstNot comes
+    int m_ops = 0;  // Number of operations such as And, Or, Xor, Sel...
+    int m_lsb = 0;  // Current LSB
+    LeafInfo* m_leafp = nullptr;  // AstConst or AstVarRef that currently looking for
+    AstNode* m_rootp;  // Root of this AST subtree
+    AstNode* m_curOpp = nullptr;  // The node that should be added to m_frozenNodes
+
+    AstUser4InUse m_inuser4;
+    std::vector<AstNode*> m_frozenNodes;  // Nodes that cannot be optimized
+    std::vector<VarInfo*> m_varInfo;
+
+    // NODE STATE
+    // AstVarRef::user4u      -> Base index of m_varInfo that points VarInfo
+    // AstVarScope::user4u    -> Same as AstVarRef::user4
+
+    // METHODS
+    VL_DEBUG_FUNC;  // Declare debug()
+
+    bool isAndTree() const { return VN_IS(m_rootp, And); }
+    bool isOrTree() const { return VN_IS(m_rootp, Or); }
+    bool isXorTree() const { return VN_IS(m_rootp, Xor) || VN_IS(m_rootp, RedXor); }
+
+#define CONST_BITOP_RETURN_IF(cond, nodep) \
+    if (setFailed(cond, #cond, nodep, __LINE__)) return
+#define CONST_BITOP_SET_FAILED(reason, nodep) setFailed(true, reason, nodep, __LINE__)
+    bool setFailed(bool fail, const char* reason, AstNode* nodep, int line) {
+        if (fail) {
+            UINFO(9, m_rootp << " cannot be optimized. reason:" << reason << " called from line:"
```

Comment:
```suggestion
            UINFO(9, "cannot optimize "<<m_rootp<<" reason:" << reason << " called from line:"
```

(Prefer messages with some text first as I sometimes sort messages)